iothub/device/src/Transport/Mqtt/OrderedTwoPhaseWorkQueue.cs

```diff
@@ -15,71 +16,78 @@ private class IncompleteWorkItem
         {
             public IncompleteWorkItem(TWorkId id, TWork workItem)
             {
-                this.WorkItem = workItem;
-                this.Id = id;
+                WorkItem = workItem ?? throw new ArgumentNullException(nameof(workItem));
+                Id = id ?? throw new ArgumentNullException(nameof(id));
             }
 
             public TWork WorkItem { get; }
 
             public TWorkId Id { get; }
         }
 
-        private readonly Func<TWork, TWorkId> getWorkId;
-        private readonly Func<IChannelHandlerContext, TWork, Task> completeWork;
-        private readonly Queue<IncompleteWorkItem> incompleteQueue = new Queue<IncompleteWorkItem>();
+        private readonly Func<TWork, TWorkId> _getWorkId;
+        private readonly Func<IChannelHandlerContext, TWork, Task> _completeWorkAsync;
+        private readonly Queue<IncompleteWorkItem> _incompleteQueue = new Queue<IncompleteWorkItem>();
 
-        public OrderedTwoPhaseWorkQueue(Func<IChannelHandlerContext, TWork, Task> worker, Func<TWork, TWorkId> getWorkId, Func<IChannelHandlerContext, TWork, Task> completeWork)
-            : base(worker)
+        public OrderedTwoPhaseWorkQueue(
+            Func<IChannelHandlerContext, TWork, Task> workerAsync,
+            Func<TWork, TWorkId> getWorkId,
+            Func<IChannelHandlerContext, TWork, Task> completeWorkAsync)
+            : base(workerAsync)
         {
-            this.getWorkId = getWorkId;
-            this.completeWork = completeWork;
+            _getWorkId = getWorkId ?? throw new ArgumentNullException(nameof(getWorkId));
+            _completeWorkAsync = completeWorkAsync ?? throw new ArgumentNullException(nameof(completeWorkAsync));
         }
 
         public Task CompleteWorkAsync(IChannelHandlerContext context, TWorkId workId)
         {
-            if (this.incompleteQueue.Count == 0)
+            if (!_incompleteQueue.Any())
```

Comment:
Okay, so `Peek` and `Dequeue` will throw a `InvalidOperationException` if the queue is empty, not return null as I expected. So that can't be a source of the customer-reported NRE.

However, we have this check about there being items in the queue before we peek or dequeue, so unless multiple threads are working against it, there won't be a race condition or a chance at that exception.

I toyed with adding try/catch around it to guard against that, but as it isn't the exception the customer saw, I don't see a reason to add that protection. Again, more insight into the greater context of this class should be known before we muck with it too much.

Reply:
I thought I put the InvalidOp exception in one of our conversations. Anyway, it's correct that there won't be a race if it's single threaded.